packages/puppeteer-core/src/bidi/BrowserConnector.ts

```diff
@@ -0,0 +1,125 @@
+import {debugError} from '../common/util.js';
+import {isNode} from '../environment.js';
+
+import type {BidiBrowser} from './Browser.js';
+import type {BidiConnection} from './Connection.js';
+
+const DEFAULT_VIEWPORT = Object.freeze({width: 800, height: 600});
+
+const getWebSocketTransportClass = async () => {
+  return isNode
+    ? (await import('../node/NodeWebSocketTransport.js')).NodeWebSocketTransport
+    : (await import('../common/BrowserWebSocketTransport.js'))
+        .BrowserWebSocketTransport;
+};
+
+/**
+ * Users should never call this directly; it's called when calling `puppeteer.connect`
+ * with `protocol: 'webDriverBiDi'`. This method attaches Puppeteer to an existing browser
+ * instance. First it tries to connect to the browser using pure BiDi. If the protocol is
+ * not supported, connects to the browser using BiDi over CDP.
+ *
+ * @internal
+ */
+export async function _connectToBiDiBrowser(
+  options: BrowserConnectOptions & ConnectOptions
+): Promise<BidiBrowser> {
+  const {ignoreHTTPSErrors = false, defaultViewport = DEFAULT_VIEWPORT} =
+    options;
+
+  const bidiConnection = await getBiDiConnection(options);
+  const BiDi = await import(/* webpackIgnore: true */ './bidi.js');
+  const bidiBrowser = await BiDi.BidiBrowser.create({
+    connection: bidiConnection,
+    closeCallback: async () => {
+      bidiConnection.send('browser.close', {}).catch(debugError);
+    },
+    process: undefined,
+    defaultViewport: defaultViewport,
+    ignoreHTTPSErrors: ignoreHTTPSErrors,
+  });
+  return bidiBrowser;
+}
+
+/**
+ * Returns a BiDiConnection established to the endpoint specified by the options.
+ * The method tries to connect to the browser using pure BiDi protocol, and falls back
+ * to BiDi over CDP if it fails.
+ */
+async function getBiDiConnection(
+  options: BrowserConnectOptions & ConnectOptions
+): Promise<BidiConnection> {
+  const {
+    ignoreHTTPSErrors = false,
+    browserWSEndpoint,
+    slowMo = 0,
+    protocolTimeout,
+  } = options;
+
+  // Try pure BiDi first.
+  if (browserWSEndpoint !== undefined) {
+    const WebSocketClass = await getWebSocketTransportClass();
+    const wsTransport: ConnectionTransport =
+      await WebSocketClass.create(browserWSEndpoint);
+    const BiDi = await import(/* webpackIgnore: true */ './bidi.js');
+    const pureBidiConnection = new BiDi.BidiConnection(
+      browserWSEndpoint,
+      wsTransport,
+      slowMo,
+      protocolTimeout
+    );
+    try {
+      const result = await pureBidiConnection.send('session.status', {});
+      if ('type' in result && result.type === 'success') {
+        // The `browserWSEndpoint` points to an endpoint supporting pure WebDriver BiDi.
+        return pureBidiConnection;
+      }
+    } catch (e: any) {
+      if (!('name' in e && e.name === 'ProtocolError')) {
+        // Unexpected exception not related to BiDi / CDP. Rethrow.
+        throw e;
+      }
+      wsTransport.close();
```

Comment:
Currently this only closes the `wsTransport` if it's expected error. I think we need to close it if any error occurrences.

Reply:
if the error is unexpected, we don't continue execution, so I don't think we need to clean up the state in such a case.